File: Backend/src/main/java/com/juan/estevez/sistemapagos/services/PagoService.java
package com.juan.estevez.sistemapagos.services;

import com.juan.estevez.sistemapagos.dto.PagoRequest;
import com.juan.estevez.sistemapagos.dto.PagoResponse;
import com.juan.estevez.sistemapagos.enums.EstadoPago;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.math.BigDecimal;
import java.time.LocalDate;

public interface PagoService {

    PagoResponse crearPago(PagoRequest request, MultipartFile archivo, String username);

    PagoResponse actualizarEstado(Long id, EstadoPago estado, String observaciones, String username);

    PagoResponse obtenerPago(Long id);

    Page<PagoResponse> listarPagos(Pageable pageable);

    Page<PagoResponse> buscarPagos(String codigoEstudiante, EstadoPago estado,
            String tipo, LocalDate fechaDesde,
            LocalDate fechaHasta, Pageable pageable);

    Page<PagoResponse> listarPagosPorEstudiante(String codigo, Pageable pageable);

    byte[] descargarArchivo(Long id);

    BigDecimal obtenerTotalPagado(String codigoEstudiante);
}

File: sistema-pagos-frontend/src/app/core/services/pago.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { EstadoPago, PagoRequest, PagoResponse, TipoPago } from '../models/pago.model';
import { Observable } from 'rxjs';
import { PaginatedResponse } from './estudiante.service';

@Injectable({
  providedIn: 'root',
})
export class PagoService {
  
  constructor(private apiService: ApiService) { }

  crearPago(pagoRequest: PagoRequest): Observable<PagoResponse> {
    const formData = new FormData();

    formData.append('data', JSON.stringify({
      cantidad: pagoRequest.cantidad,
      tipoPago: pagoRequest.tipoPago,
      fecha: pagoRequest.fecha,
      codigoEstudiante: pagoRequest.codigoEstudiante,
      observaciones: pagoRequest.observaciones
    }));

    formData.append('archivo', pagoRequest.archivo);
    
    return this.apiService.postMultipart<PagoResponse>('/pagos', formData);
  }

  obtenerPago(id: number): Observable<PagoResponse> {
    return this.apiService.get<PagoResponse>(`/pagos/${id}`);
  }

  listarPagos(
    page: number = 0,
    size: number = 10
  ): Observable<PaginatedResponse<PagoResponse>> {
    return this.apiService.get<PaginatedResponse<PagoResponse>>('/pagos', {
      page,
      size,
      sort: 'fecha,desc'
    });
  }

  buscarPagos(
    codigoEstudiante?: string,
    estado?: EstadoPago,
    tipo?: TipoPago,
    fechaDesde?: Date,
    fechaHasta?: Date,
    page: number = 0,
    size: number = 10
  ): Observable<PaginatedResponse<PagoResponse>> {
    const params: any = { page, size };
    if (codigoEstudiante) params.codigoEstudiante = codigoEstudiante;
    if (estado) params.estado = estado;
    if (tipo) params.tipo = tipo;
    if (fechaDesde) params.fechaDesde = fechaDesde;
    if (fechaHasta) params.fechaHasta = fechaHasta;

    return this.apiService.get<PaginatedResponse<PagoResponse>>('/pagos/buscar', params);
  }

  listarPagosPorEstudiante(
    codigo: string,
    page: number = 0,
    size: number = 10
  ): Observable<PaginatedResponse<PagoResponse>> {
    return this.apiService.get<PaginatedResponse<PagoResponse>>(`/pagos/estudiante/${codigo}`, {
      page,
      size
    });
  }

  actualizarEstado(
    id: number,
    estado: EstadoPago,
    observaciones?: string
  ): Observable<PagoResponse> {
    return this.apiService.patch<PagoResponse>(`/pagos/${id}/estado`, {
      estado,
      observaciones
    });
  }

  descargarArchivo(id: number): Observable<Blob> {
    return this.apiService.downloadFile(`/pagos/${id}/archivo`);
  }

  obtenerTotalPagado(codigoEstudiante: string): Observable<number> {
    return this.apiService.get<number>(`/pagos/estudiante/${codigoEstudiante}/total`);
  }

}
